Add S3 storage and multi-hop transfer tutorial section (#766)
rucio/documentation: docs/operator/setting_up_demo.md

| test  | file4 | 10.486 MB | 65786e49 | XRD3: root://xrd3:1096//rucio/... |
+-------+-------+-----------+----------+-----------------------------------+
```

# Configuring S3 Storage and Multi-Hop Transfers in Rucio

This tutorial covers how to register S3-compatible storage (MinIO) as Rucio Storage Elements (RSEs), configure credentials for both Rucio and FTS, and set up RSE distances to enable multi-hop transfers between S3 and XRootD endpoints.

> The examples use a Docker Compose playground environment with two MinIO instances (MINIO1, MINIO2) and three XRootD servers (XRD1, XRD2, XRD3). The commands assume you are already operating within a Rucio admin environment with the `rucio` and `rucio-admin` CLI tools available.

Reviewer: For clarity, `rucio` and `rucio-admin` have been merged since rucio 38, the `admin` role is handled by the permission policies.  Instead I'd change this too 
```suggestion
The examples use a Docker Compose playground environment with two MinIO instances (MINIO1, MINIO2) and three XRootD servers (XRD1, XRD2, XRD3). The commands assume you are already have an rucio instance with an admin account. 
```

```bash
rucio rse protocol add XRD3 \
  --host xrd3 \
  --scheme https \
  --prefix //rucio \
  --port 1096 \
  --impl rucio.rse.protocols.gfal.Default \
  --domain-json '{"wan": {"read": 2, "write": 2, "delete": 2, "third_party_copy_read": 2, "third_party_copy_write": 2}, "lan": {"read": 2, "write": 2, "delete": 2}}'
```

The priority values (`"read": 2` etc.) ensure that the existing XRootD protocol remains preferred for direct transfers, while HTTPS is available for multi-hop routing.

## Creating Buckets on MinIO

Before registering MinIO instances as RSEs, create the `rucio` bucket on each. This uses the MinIO Client (`mc`) from within each MinIO container:

```bash
# On MINIO1
export MC_INSECURE=true
mc alias set local https://localhost:9001 admin password
mc mb local/rucio

# On MINIO2
export MC_INSECURE=true
mc alias set local https://localhost:9002 admin password
mc mb local/rucio
```

## Registering MinIO RSEs

Register both MinIO instances as RSEs with S3 protocol configuration. The `gfal.NoRename` implementation is used because S3 does not support server-side rename operations.

> ```bash

Reviewer: This can be reduced to a for loop (over both MINIO1 and 2), it would make this easier to read

rucio rse add MINIO1
rucio rse protocol add MINIO1 \
  --host minio1 \
  --port 9001 \
  --scheme https \
  --prefix /rucio/ \
  --impl rucio.rse.protocols.gfal.NoRename \
  --domain-json '{"lan": {"read": 1, "write": 1, "delete": 1}, "wan": {"read": 1, "write": 1, "delete": 1, "third_party_copy_read": 1, "third_party_copy_write": 1}}'

> rucio rse attribute add MINIO1 --key sign_url --value s3

Reviewer: These could use some inline comments to explain each of the options, or at least a link to the config params page in the section description

rucio rse attribute add MINIO1 --key fts --value https://fts:8446
rucio account limit add root --rse MINIO1 --bytes infinity

rucio rse add MINIO2
rucio rse protocol add MINIO2 \
  --host minio2 \
  --port 9002 \
  --scheme https \
  --prefix /rucio/ \
  --impl rucio.rse.protocols.gfal.NoRename \
  --domain-json '{"lan": {"read": 1, "write": 1, "delete": 1}, "wan": {"read": 1, "write": 1, "delete": 1, "third_party_copy_read": 1, "third_party_copy_write": 1}}'
rucio rse attribute add MINIO2 --key sign_url --value s3
rucio rse attribute add MINIO2 --key s3_url_style --value path
rucio rse attribute add MINIO2 --key verify_checksum --value False
rucio rse attribute add MINIO2 --key skip_upload_stat --value True
rucio rse attribute add MINIO2 --key strict_copy --value True
rucio rse attribute add MINIO2 --key fts --value https://fts:8446
rucio account limit add root --rse MINIO2 --bytes infinity
```

### Setting RSE Credentials

Rucio needs S3 credentials to generate presigned URLs for transfers. These are stored in `rse-accounts.cfg`, keyed by RSE ID:

```bash
ID1=$(rucio rse show MINIO1 | grep '^  id:' | awk '{print$2}')
ID2=$(rucio rse show MINIO2 | grep '^  id:' | awk '{print$2}')

> cat >/opt/rucio/etc/rse-accounts.cfg <<JSON

Reviewer: This doesn't render to anything special, I'd instead just use the `console` tag to instead do

```console
$ cat /opt/rucio/etc/rse-accounts.cfg
{
  "$ID1": {
    "access_key": "admin",
    "secret_key": "password",
    "signature_version": "s3v4",
    "region": "us-east-1"
  },
  "$ID2": {
    "access_key": "admin",
    "secret_key": "password",
    "signature_version": "s3v4",
    "region": "us-east-1"
  }
}
```

{
  "$ID1": {
    "access_key": "admin",
    "secret_key": "password",
    "signature_version": "s3v4",
    "region": "us-east-1"
  },
  "$ID2": {
    "access_key": "admin",
    "secret_key": "password",
    "signature_version": "s3v4",
    "region": "us-east-1"
  }
}
JSON
```

### Configuring RSE Distances for Multi-Hop

> RSE distances tell Rucio which transfer paths are available and their relative cost. Setting a distance of 1 between MINIO RSEs and XRD3 establishes the multi-hop path: transfers from MinIO to XRD1 or XRD2 will route through XRD3 as an intermediate.

Reviewer: This explanation is technically correct, but a little overly verbose. I recommend putting in a mermaid chart instead to make this more visual. 

```suggestion
RSE distances establish transfer paths between RSEs. Setting a distance of 1 between the source and an intermediate will ensure the intermediate transfer will always be preferred over longer direct transfers.  

```mermaid
graph TD
    MINIO[MINIO RSE]
    XRD1[XRD1 RSE]
    XRD2[XRD2 RSE]
    XRD3[XRD3 RSE]
    
    MINIO -.->|distance=1| XRD3
    XRD3 -.->|distance=1| XRD1
    XRD3 -.->|distance=1| XRD2
```

```
